test/unit/sharding_key_test.lua

```diff
@@ -0,0 +1,87 @@
+local t = require('luatest')
+local sharding_key_module = require('crud.common.sharding_key')
+local utils = require('crud.common.utils')
+
+local g = t.group('sharding_key')
+
+g.before_each(function()
+    local sharding_key_format = {
+        {name = 'space_name', type = 'string', is_nullable = false},
+        {name = 'sharding_key', type = 'array', is_nullable = false}
+    }
+    -- Create a space _ddl_sharding_key with a tuple that
+    -- contains a space name and it's sharding key.
+    if type(box.cfg) ~= 'table' then
+        box.cfg{}
+    end
+    box.schema.space.create('_ddl_sharding_key', {
+        format = sharding_key_format,
+        if_not_exists = true
+    })
+    box.space._ddl_sharding_key:create_index('pk', {if_not_exists = true})
+end)
+
+g.after_each(function()
+    -- Cleanup.
+    box.space._ddl_sharding_key:truncate()
+    box.space._ddl_sharding_key:drop()
+end)
+
+g.test_as_index_object_positive = function()
+    local space_name = 'as_index_object'
+    local space_format = {
+        {name = 'id', type = 'unsigned'},
+        {name = 'name', type = 'string'},
+        {name = 'age', type = 'unsigned'},
+    }
+    local sharding_key_def = {'name', 'age'}
+
+    local index_obj, err = sharding_key_module.as_index_object(space_format,
+                                                               space_name,
+                                                               sharding_key_def)
+    t.assert_equals(err, nil)
+    t.assert_equals(index_obj, {
+        parts = {
+            {fieldno = 2},
+            {fieldno = 3},
+        }
+    })
+end
+
+g.test_as_index_object_negative = function()
+    local space_name = 'as_index_object'
+    local space_format = {
+        {name = 'id', type = 'unsigned'},
+        {name = 'name', type = 'string'},
+        {name = 'age', type = 'unsigned'},
+    }
+    local sharding_key_def = {'dude', 'age'}
+
+    local index_obj, err = sharding_key_module.as_index_object(space_format,
+                                                               space_name,
+                                                               sharding_key_def)
+    t.assert_str_contains(err.err,
+        'No such field (dude) in a space format (as_index_object)')
+    t.assert_equals(index_obj, nil)
+end
+
+g.test_get_format_fieldno_map = function()
+    local space_format = {
+        {name = 'id', type = 'unsigned'},
+        {name = 'name', type = 'string'},
+        {name = 'age', type = 'unsigned'},
+    }
+
+    local fieldno_map = utils.get_format_fieldno_map(space_format)
+    t.assert_equals(fieldno_map, {age = 3, id = 1, name = 2})
+end
+
+g.test_fetch_on_storage = function()
+    local space_name = 'fetch_on_storage'
+    local sharding_key_def = {'name', 'age'}
+    box.space._ddl_sharding_key:upsert(
+        {space_name, sharding_key_def}, {{'=', 1, space_name}})
+
+    local key = sharding_key_module.fetch_on_storage()
```

Comment:
Nit: `key` is the map of all sharding keys across all spaces? The naming looks confusing.

Reply:
Yes, it's a map.

```lua
--- a/test/unit/sharding_key_test.lua
+++ b/test/unit/sharding_key_test.lua
@@ -82,8 +82,8 @@ g.test_fetch_on_storage = function()
     box.space._ddl_sharding_key:upsert(
         {space_name, sharding_key_def}, {{'=', 1, space_name}})
 
-    local key = sharding_key_module.fetch_on_storage()
-    t.assert_equals(key, {[space_name] = sharding_key_def})
+    local map_with_keys = sharding_key_module.fetch_on_storage()
+    t.assert_equals(map_with_keys, {[space_name] = sharding_key_def})
 end
 
 g.test_extract_from_index = function()
```

```lua
--- a/crud/common/sharding_key.lua
+++ b/crud/common/sharding_key.lua
@@ -71,11 +71,11 @@ local function reentrant_fetch(replicaset)
                         'Waiting for sharding key to be fetched is timed out')
     end
 
-    local sharding_key_def = {}
+    local map_with_keys = {}
     if sharding_key_fetch_in_progress == false then
         sharding_key_fetch_in_progress = true
         local err
-        sharding_key_def, err = replicaset:call(FETCH_FUNC_NAME, {},
+        map_with_keys, err = replicaset:call(FETCH_FUNC_NAME, {},
                                                 {timeout = const.FETCH_SHARDING_KEY_TIMEOUT})
         if err ~= nil then
             sharding_key_fetch_in_progress = false
@@ -85,7 +85,7 @@ local function reentrant_fetch(replicaset)
         sharding_key_fetch_in_progress = false
     end
 
-    return sharding_key_def
+    return map_with_keys
 end
 
 -- Get sharding key definitions (field names) for all spaces or a certain space
```

Reply:
Heh, using of `sharding_key_def` as for the key definition as well as for the mapping from a space name to the key definition is one of nitpickings I plan to leave soon :) I would also add that usage of those terms should be consistent across the project.

Regarding the `map_with_keys` name: those are not keys, those are key definitions. `sharding_key_def_map` is long. Can we call it, say, `sharding_key_dd` ([data dictionary](https://en.wikipedia.org/wiki/Data_dictionary))? I'm not sure here, so I'll leave this comment to think more.

Reply:
I like `sharding_key_def_map` or `key_def_map` (this variable used in 10 lines, so I believe we can omit prefix `sharding`).

Reply:
changed to `sharding_key_def_map`